### ISSUE TITLE
Dynamic Inventories now fail if one or more VMs do not have a public IP address

### ISSUE THREAD

**Issue by epopisces** [ansible/ansible_collections_azure#1378]
##### SUMMARY
We use AWX, version 23.5.1 and use the azure dynamic inventory feature.  Up until this week performing the sync operation on an inventory source ran without issue.

[Pull Request 1214](https://github.com/ansible-collections/azure/pull/1214) appears to have introduced a bug: a 'list index out of range' error halts execution when the inventory sync process encounters a VM that does not have a public IP (which is the vast majority of our environments: we use Azure Bastion as the sole mechanism of access for most of our VMs).  This resulted in nearly all of our dynamic inventories breaking and our hosts being unable to be managed by Ansible via AWX without manual work.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
Azure Dynamic Inventory/Inventory > Sources > Sync process

##### ANSIBLE VERSION

```paste below
ansible [core 2.15.8]
  config file = None
  configured module search path = ['/runner/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible
  ansible collection location = /runner/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.9.18 (main, Sep  7 2023, 00:00:00) [GCC 11.4.1 20230605 (Red Hat 11.4.1-2)] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION

```paste below
Collection         Version
------------------ -------
azure.azcollection 2.1.0
```

##### CONFIGURATION

```paste below
CONFIG_FILE() = None
```

##### OS / ENVIRONMENT

AWX 23.5.1 running in AKS, deployed via the latest awx-operator (version 2.9.0).  Issue occurs when using the latest Ansible-provided Execution Environment (https://quay.io/repository/ansible/awx-ee).

##### STEPS TO REPRODUCE
Set up an AWX environment with Azure SPN, and a azure_rm.yml file (named all_running_hosts.azure_rm.yml in our case) with contents like yaml shown below.  Create an Inventory with a Sources entry that uses that file (see screenshot for example). 

```yaml
---
plugin: azure.azcollection.azure_rm
auth_source: auto
plain_host_names: true
# unpublished Azure API batching fails when over 10 VMs in a subscription
batch_fetch: false
exclude_host_filters:
  - powerstate != 'running'
keyed_groups:
  - prefix: tag
    key: tags
  - prefix: loc
    key: location
  - prefix: os
    key: os_disk.operating_system_type
conditional_groups:
  all_hosts: true
  staging: "'stg' in name"
  production: "'prod' or 'prd' in name"
  development: "'dev' in name"
hostvar_expressions:
  ansible_host: private_ipv4_addresses | first

#* References
# guide: https://faun.pub/lets-do-devops-dynamic-host-inventories-in-azure-on-ansible-awx-tower-87437f2838c1 with source
# from https://github.com/KyMidd/AnsibleAwxAzureDynamicInventory/blob/master/inventories/azure/all_running_hosts.azure_rm.yml
# incredibly helpful in building out keyed_groups: https://www.azurecitadel.com/packeransible/dynamic/
# https://docs.ansible.com/ansible/latest/collections/azure/azcollection/azure_rm_inventory.html

```
![image](https://github.com/ansible-collections/azure/assets/17202697/ee4718a6-0b64-46a3-995c-79edafc78984)

##### EXPECTED RESULTS

Expect a successful inventory source sync process dynamically populating our inventory with hosts.

##### ACTUAL RESULTS
Each sync for an environment that contains one or more running hosts that do not have a public IP address fails with the following error:
```paste below
  config file = /runner/project/ansible.cfg
  configured module search path = ['/runner/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible
  ansible collection location = /runner/requirements_collections:/runner/.ansible/collections:/usr/share/ansible/collections:/usr/share/automation-controller/collections
  executable location = /usr/local/bin/ansible-inventory
  python version = 3.9.18 (main, Sep  7 2023, 00:00:00) [GCC 11.4.1 20230605 (Red Hat 11.4.1-2)] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
Using /runner/project/ansible.cfg as config file
redirecting (type: inventory) ansible.builtin.azure_rm to azure.azcollection.azure_rm
[WARNING]:  * Failed to parse
/runner/project/inventories/azure/all_running_hosts.azure_rm.yml with
ansible_collections.azure.azcollection.plugins.inventory.azure_rm plugin: list
index out of range
  File "/usr/local/lib/python3.9/site-packages/ansible/inventory/manager.py", line 293, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/runner/requirements_collections/ansible_collections/azure/azcollection/plugins/inventory/azure_rm.py", line 232, in parse
    self._get_hosts()
  File "/runner/requirements_collections/ansible_collections/azure/azcollection/plugins/inventory/azure_rm.py", line 319, in _get_hosts
    next(chain([h.hostvars['public_ip_address'][0]['ipv4_address']], h.hostvars['private_ipv4_addresses']), None))
[WARNING]: Unable to parse
/runner/project/inventories/azure/all_running_hosts.azure_rm.yml as an
inventory source
ERROR! No inventory was parsed, please check your configuration and options.
```

Note that a given credential (as used by a Inventory Source entry) for us is scoped to a single Azure subscription, so we have multiple Sources for our Inventory, one for each subscription AWX/Ansible managed.  This issue impacts every subscription with a VM that does not have a public IP (which is the vast majority of them).

See also discussion [on the merge itself](https://github.com/ansible-collections/azure/pull/1214)

```

**Comment by emanueleaina** [comment]
In my inventory I was relying on the default `ansible_host` handling for the VMs that do not have a public IP address, but now:

* with older versions I get the exception described in https://github.com/ansible-collections/azure/issues/1328
  * can be worked around by setting `batch_fetch: false`
* with 2.1.0 I get the `chain()` exception above
* with 2.1.1 `ansible_host` is being set to an empty list (`[]`)

**Comment by Fred-sun** [comment]
Fixes in #1379
